drivers/input/input_sbus.c

```diff
@@ -0,0 +1,364 @@
+/*
+ * Copyright (c) 2024 CogniPilot Foundation
+ * Copyright (c) 2024 NXP Semiconductors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT futaba_sbus
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/input/input.h>
+#include <zephyr/irq.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/time_units.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/drivers/uart.h>
+
+LOG_MODULE_REGISTER(futaba_sbus, CONFIG_INPUT_LOG_LEVEL);
+
+/* Driver config */
+struct input_channel_config {
+	uint32_t channel;
+	uint32_t type;
+	uint32_t zephyr_code;
+};
+
+const struct uart_config uart_cfg_sbus = {.baudrate = 100000,
+					  .parity = UART_CFG_PARITY_EVEN,
+					  .stop_bits = UART_CFG_STOP_BITS_2,
+					  .data_bits = UART_CFG_DATA_BITS_8,
+					  .flow_ctrl = UART_CFG_FLOW_CTRL_NONE};
+
+struct input_sbus_config {
+	uint8_t num_channels;
+	const struct input_channel_config *channel_info;
+	const struct device *uart_dev;
+	uart_irq_callback_user_data_t cb;
+};
+
+#define SBUS_FRAME_LEN 25
+#define SBUS_HEADER    0x0F
+#define SBUS_FOOTER    0x00
+
+#define SBUS_BYTE24_CH17       0x01
+#define SBUS_BYTE24_CH18       0x02
+#define SBUS_BYTE24_FRAME_LOST 0x04
+#define SBUS_BYTE24_FAILSAFE   0x08
+
+#define SBUS_TRANSMISSION_TIME  4  /* ms */
+#define SBUS_INTERFRAME_SPACING 20 /* ms */
+#define SBUS_CHANNEL_COUNT      16
+
+#define REPORT_FILTER      CONFIG_INPUT_SBUS_REPORT_FILTER
+#define CHANNEL_VALUE_ZERO CONFIG_INPUT_SBUS_CHANNEL_VALUE_ZERO
+#define CHANNEL_VALUE_ONE  CONFIG_INPUT_SBUS_CHANNEL_VALUE_ONE
+
+struct input_sbus_data {
+	struct k_thread thread;
+	struct k_sem report_lock;
+
+	uint16_t xfer_bytes;
+	uint8_t rd_data[SBUS_FRAME_LEN];
+	uint8_t sbus_frame[SBUS_FRAME_LEN];
+	bool partial_sync;
+	bool in_sync;
+	uint32_t last_rx_time;
+
+	uint16_t last_reported_value[SBUS_CHANNEL_COUNT];
+	int8_t channel_mapping[SBUS_CHANNEL_COUNT];
+
+	K_KERNEL_STACK_MEMBER(thread_stack, CONFIG_INPUT_SBUS_THREAD_STACK_SIZE);
+};
+
+static void input_sbus_report(const struct device *dev, unsigned int sbus_channel,
+			      unsigned int value)
+{
+	const struct input_sbus_config *const config = dev->config;
+	struct input_sbus_data *const data = dev->data;
+
+	int channel = data->channel_mapping[sbus_channel];
+
+	/* Not Mapped */
+	if (channel == -1) {
+		return;
+	}
+
+	if (value > (data->last_reported_value[channel] + REPORT_FILTER) ||
+	    value < (data->last_reported_value[channel] - REPORT_FILTER)) {
+		switch (config->channel_info[channel].type) {
+		case INPUT_EV_ABS:
+		case INPUT_EV_MSC:
```

Comment:
either use a `case default` for this or add a default with an error so data don't get silently discarded if configured incorrectly

even better validate the type at build time as well